tests/test_device.py

```diff
@@ -0,0 +1,106 @@
+# SPDX-License-Identifier: MIT
+
+import pyudev
+import pytest
+
+from ratbag_emu.device import Device
+from ratbag_emu.util import EventData
+from ratbag_emu.actuators.sensor import SensorActuator
+
+from tests import TestBase
+
+
+class TestDevice(TestBase):
+    name = 'Test Device'
+    vid = pid = 0x9999
+
+    @pytest.fixture()
+    def device(self):
+        d = Device(name=self.name, info=(3, self.vid, self.pid))
+
+        yield d
+
+        d.destroy()
+
+    def test_create(self, device):
+        '''
+        Make sure the correct device is created
+        '''
+        name = f'ratbag-emu {device.id} ({self.name}, {self.vid:04x}:' \
+               f'{self.pid:04x}, 0)'
+        assert list(pyudev.Context().list_devices(subsystem='hid',
+                                                  HID_NAME=name))
+
+    def test_duplicated_actuators(self, device):
+        '''
+        Make sure we raise an error on actuators that operate on the same keys
+        '''
+        with pytest.raises(AssertionError):
+            device.actuators += [
+                SensorActuator(device, dpi=1000),
+                SensorActuator(device, dpi=1000)
+            ]
+
+    def test_mouse_report(self, device, events):
+        '''
+        Test single HID report
+        '''
+        x = y = 5
+
+        expected = EventData(x, y)
+
+        def callback(device):
+            nonlocal expected
+            device.send_hid_action(expected)
+
+        received = self.catch_events(device, events, callback)
+
+        assert expected.x <= received.x <= expected.x
+        assert expected.y <= received.y <= expected.y
+
+    def test_movement(self, device, events):
+        '''
+        Test normal mouse movement
+        '''
+        dpi = 1000
+
+        device.actuators += [
+            SensorActuator(device, dpi)
+        ]
+
+        action = {
+            'type': 'xy',
+            'duration': 500,
+            'data': {
+                'x': 5,
+                'y': 5
+            }
+        }
+
+        received = self.simulate(device, events, action)
+
+        expected = EventData.from_action(dpi, action)
+
+        assert received.x == expected.x
+        assert received.y == expected.y
+
+    def test_movement_max(self, device, events):
+        '''
+        Make sure we raise an error when we try to move more than possible
+        '''
+        device.report_rate = 1000
+        device.actuators += [
+            SensorActuator(device, dpi=1000)
+        ]
```

Comment:
I have a feeling the `SensorActuator` should be defined statically when creating the device for the test.
Actually, we might even want to have in the future a logic that parses the provided report descriptor, and creates the corresponding actuators.

Reply:
What do you mean here? Defined in the device constructor?

Populating the actuators based on the report descriptor would only make sense for generic devices.

Reply:
> What do you mean here? Defined in the device constructor?

I think so: `Device` could very well go through the report descriptors, and create the `Actuator` objects it can detect.

This comment was because all of your tests have a lot of duplicated statement for those, and part of it can be automated.

> Populating the actuators based on the report descriptor would only make sense for generic devices.

In a way, yes, but you still can pre-populate  the list of Actuators.

But honestly, while thinking about this more, it might not help as well.

[few seconds and a ahah moment later]

OK, I think I know what should be done for generic devices:
- the tests define the `Actuator`
- then while creating the device, based on the list of `Actuators`, we generate the report descriptor.

This can come in in  a later PR.

Reply:
> I think so: Device could very well go through the report descriptors, and create the Actuator objects it can detect.
>
> This comment was because all of your tests have a lot of duplicated statement for those, and part of it can be automated.

But it isn't the same statement. I just add what I need to `device.actuators`, all statements are similar but not quite the same. If I wanted the same statement in all of them I could just put it in the `device` fixture declared above.

```python
    @pytest.fixture()
    def device(self):
        d = Device(name=self.name, info=(3, self.vid, self.pid), rdescs=[[
                        # Generic mouse report descriptor
                        0x05, 0x01,  # .Usage Page (Generic Desktop)        0
                        0x09, 0x02,  # .Usage (Mouse)                       2
                        0xa1, 0x01,  # .Collection (Application)            4
                        0x09, 0x02,  # ..Usage (Mouse)                      6
                        0xa1, 0x02,  # ..Collection (Logical)               8
                        0x09, 0x01,  # ...Usage (Pointer)                   10
                        0xa1, 0x00,  # ...Collection (Physical)             12
                        0x05, 0x09,  # ....Usage Page (Button)              14
                        0x19, 0x01,  # ....Usage Minimum (1)                16
                        0x29, 0x03,  # ....Usage Maximum (3)                18
                        0x15, 0x00,  # ....Logical Minimum (0)              20
                        0x25, 0x01,  # ....Logical Maximum (1)              22
                        0x75, 0x01,  # ....Report Size (1)                  24
                        0x95, 0x03,  # ....Report Count (3)                 26
                        0x81, 0x02,  # ....Input (Data,Var,Abs)             28
                        0x75, 0x05,  # ....Report Size (5)                  30
                        0x95, 0x01,  # ....Report Count (1)                 32
                        0x81, 0x03,  # ....Input (Cnst,Var,Abs)             34
                        0x05, 0x01,  # ....Usage Page (Generic Desktop)     36
                        0x09, 0x30,  # ....Usage (X)                        38
                        0x09, 0x31,  # ....Usage (Y)                        40
                        0x15, 0x81,  # ....Logical Minimum (-127)           42
                        0x25, 0x7f,  # ....Logical Maximum (127)            44
                        0x75, 0x08,  # ....Report Size (8)                  46
                        0x95, 0x02,  # ....Report Count (2)                 48
                        0x81, 0x06,  # ....Input (Data,Var,Rel)             50
                        0xc0,        # ...End Collection                    52
                        0xc0,        # ..End Collection                     53
                        0xc0,        # .End Collection                      54
                     ]])

        yield d

        d.destroy(
```

Although your approach would be cleaner I have some doubts, see my next reply.

Also, this is what it would look like in real code.

```python
class MyCustomDevice(Device):
	def __init__(self, ...):
		...
		self.actuators = [
			CustomActuator1(...)
			CustomActuator2(...)
			CustomActuator3(...)
			CustomActuator4(...)
			CustomActuator5(...)
			CustomActuator6(...)
			CustomActuator7(...)
			CustomActuator8(...)
		]
```

And in a test class where we use the same actuator for all devices.

```python
class TestSomething(TestBase):
	@pytest.fixture()
    def device(self):
        d = Device(name=..., info=(...), rdescs=[...])
		d.actuators = [
			CustomActuator1(...)
			CustomActuator2(...)
			CustomActuator3(...)
			CustomActuator4(...)
		]

        yield d

        d.destroy(
```

Or we would just declare a custom device for it.

```python
class TestSomethingElse(TestBase):
	@pytest.fixture()
    def device(self):
        d = MyCustomDevice(...)

        yield d

        d.destroy(
```

The approach would depend on the problem. Honestly, I don't really see us needing the 1st approach (declare the actuators in the `device` fixture).

> In a way, yes, but you still can pre-populate the list of Actuators.
>
> But honestly, while thinking about this more, it might not help as well.
>
> [few seconds and a ahah moment later]
>
> OK, I think I know what should be done for generic devices:
>
>    - the tests define the Actuator
>    - then while creating the device, based on the list of Actuators, we generate the report descriptor.
>
> This can come in in a later PR.

That does make sense, however I have something to ask: which problem are we trying to solve here? Don't get me wrong, that's not a bad idea, it's just that it's not that simple to implement and I get lost on the motivation behind it. The motivation behind ratbag-emu is for us to be able to emulate real devices and for that this is completely irrelevant.

I think automated actuator discovery would be cool to add later, but at this point I don't really see the need. Do you agree?